FileStorageLevels/ReadLeveledStorage.cpp

```diff
@@ -0,0 +1,309 @@
+#include "Storage.h"
+#include <boost/archive/binary_iarchive.hpp>
+#include <boost/archive/binary_oarchive.hpp>
+#include <boost/lexical_cast.hpp>
+#include <boost/serialization/set.hpp>
+#include <boost/uuid/uuid_io.hpp>
+#include <boost/uuid/uuid_serialize.hpp>
+#include <chrono>
+#include <cstring>
+#include <fcntl.h>
+#include <fstream>
+#include <iostream>
+#include <unistd.h>
+
+namespace fs = std::filesystem;
+using namespace std::chrono_literals;
+
+ReadLeveledStorage::ReadLeveledStorage(const Storage& first_, const Storage& second_)
+{
+    extfile_name = first_.path;
+    extfile_name.replace_extension("extents");
+
+    if(!fs::exists(first_.path) || !loadExtentFile() || signature != boost::lexical_cast<boost::uuids::uuid>(sig_str))
+    {
+        // TODO: log problem
+        std::error_code ec;
+        fs::remove(extfile_name, ec);
+        fs::remove(first_.path, ec);
+        extents.clear();
+        extents.emplace(Extent{.offset = 0, .length = 0});
+        is_fully_primed = false;
+        primed_file_size = 0;
+        signature = boost::lexical_cast<boost::uuids::uuid>(sig_str);
+    }
+    file_size = fs::file_size(second_.path);
+    first_fd = open(first_.path.c_str(), O_RDWR | O_CREAT, S_IRUSR | S_IWUSR);
+    checkAndThrow(first_fd);
+    second_fd = open(second_.path.c_str(), O_RDONLY);
+    checkAndThrow(second_fd);
+    auto res = ftruncate(first_fd, file_size);
+    checkAndThrow(res);
+    extent_thread = std::thread([this]() {
+        while(!shutting_down && !is_fully_primed)
+        {
+            std::this_thread::sleep_for(1s);
+            saveExtentFile();
+        }
+    });
+}
+
+ReadLeveledStorage::~ReadLeveledStorage()
+{
+    try
+    {
+        shutting_down = true;
+        if(extent_thread.joinable())
+        {
+            extent_thread.join();
+        }
+        saveExtentFile();
+    }
+    catch(...)
+    {
+        // TODO: log problem
+    }
+}
+
+ssize_t ReadLeveledStorage::read(void* buff, size_t buff_size)
+{
+    if(buff_size == 0)
+    {
+        return 0;
+    }
+    if(is_fully_primed)
+    {
+        ++stats.primary_reads;
+        return ::read(first_fd, buff, buff_size);
+    }
+
+    ssize_t ret_val = 0;
+    while(buff_size > 0)
+    {
+        auto extent = extents.lower_bound(position);
```

Comment:
I was wondering if it was possible to search extents only once in the first loop and later on just iterate over the extents (Iterate over the tree instead of searching each extent from root)?

Reply:
it was the initial version. however, merging adjacent extents invalidates iterators, hence the lower_bound moved into the loop

Reply:
::update could somehow return the new updated iterator after the merging?

Reply:
It is interesting idea, will think about that